src/client/async_forward_backward.py

```diff
@@ -0,0 +1,215 @@
+    forward_inputs = (inputs, kwargs)
+
+    if not nested_compare(forward_inputs, rpc_info["forward_schema"]):
+        raise TypeError(f"Inputs do not match expert input schema. Did you pass the right number of parameters?")
+ 
+    forward_inputs = nested_flatten(forward_inputs)
+    inputs = tuple(tensor.cpu().detach() for tensor in forward_inputs)
+
+    serialized_tensors = (
+        serialize_torch_tensor(tensor, proto.compression)
+        for tensor, proto in zip(inputs, nested_flatten(rpc_info["forward_schema"]))
+    )
+    deserialized_outputs = await expert_forward(uid, inputs, serialized_tensors, stub)
+    flat_outputs = tuple(deserialized_outputs)
+
+    return nested_pack(flat_outputs, structure=rpc_info["outputs_schema"])
+
+
+async def run_backward(
+    uid: ModuleUID, 
+    stub: StubBase,
+    rpc_info: RPCInfo,
+    intemediate_inputs: List[torch.Tensor], 
+    grad_outputs: List[torch.Tensor], 
+) -> Sequence[torch.Tensor]:
+
+    grad_outputs_cpu = tuple(tensor.cpu() for tensor in grad_outputs)
+    inputs_and_grad_outputs = tuple(nested_flatten((intemediate_inputs, grad_outputs_cpu)))
+    backward_schema = tuple(nested_flatten((rpc_info["forward_schema"], rpc_info["outputs_schema"])))
+
+    serialized_tensors = (
+        serialize_torch_tensor(tensor, proto.compression)
+        for tensor, proto in zip(inputs_and_grad_outputs, backward_schema)
+    )
+    deserialized_grad_inputs = await expert_backward(uid, inputs_and_grad_outputs, serialized_tensors, stub)
+    return deserialized_grad_inputs
+
+
+async def async_forward(
+    inputs: torch.Tensor, 
+    sequence_manager: RemoteSequenceManager
+    ) -> Tuple[torch.Tensor, Sequence[torch.Tensor], Sequence[RemoteSpanInfo]]:
+
+    assert isinstance(inputs, torch.Tensor) and inputs.ndim == 3
+    sequences = sequence_manager.make_sequence()
+    intermediate_inputs = []
+    done_sequences = []
+
+    while len(sequences) > 0:
+        while True:
+            try:
+                span = sequences.pop(0)
+                span_uids: str = CHAIN_DELIMITER.join(sequence_manager.block_uids[span.start: span.end])
+                stub = TransformerConnectionHandler.get_stub(sequence_manager.p2p, span.peer_id)
+                (outputs, ) = await run_forward(span_uids, stub, sequence_manager.rpc_info, inputs)
+
+                assert isinstance(outputs, torch.Tensor)
+                assert outputs.shape == inputs.shape, f"Expected output {inputs.shape}, got {outputs.shape}"
+
+                # Save intermediate inputs and subsequences if the forward is already done for them
+                intermediate_inputs.append(inputs)
+                done_sequences.append(span)
+
+                inputs = outputs
+                break
+            except Exception as e:
+                logging.warning(f"Caught {e} when running forward for chain {span.start}-{span.end}", exc_info=True)
+                backup_sequences = sequence_manager[span.start: span.end].make_sequence()
```

Comment:
minor: consider make_sequence instead of lice

conceptual: consider making sequence from span_start to GLOBAL END (len(sequence_manager))
and replace sequence manager with this